### PR TITLE
Support UDF in syntax

### PR DESCRIPTION
**Pull request by imay**
Now, you can create UDF with CREATE FUNCTION statement. Doris only
support UDF in this version, it will support UDAF/UDTF later.

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/catalog/Function.java [hunk 1]
```diff
@@ -439,4 +500,160 @@ public static String getUdfType(PrimitiveType t) {
...
+
+        public void write(DataOutput output) throws IOException {
+            output.writeInt(code);
+        }
+        public static FunctionType read(DataInput input) throws IOException {
+            return fromCode(input.readInt());
+        }
+    };
+
+    protected void writeOptionString(DataOutput output, String value) throws IOException {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/catalog/Function.java]: move writeOptionString() and readOptionStringOrNull() to common.util?

File: fe/src/main/java/org/apache/doris/analysis/CreateFunctionStmt.java [hunk 2]
```diff
@@ -17,38 +17,129 @@
...
+    private final FunctionArgsDef argsDef;
+    private final TypeDef returnType;
+    private final Map<String, String> properties;
+
+    // needed item set after analyzed
+    private String objectFile;
+    private Function function;
+    private String checksum;
+
+    public CreateFunctionStmt(boolean isAggregate, FunctionName functionName, FunctionArgsDef argsDef,
```
[the remark this answers is not in the document]
Comment (morningman) [reply fe/src/main/java/org/apache/doris/analysis/CreateFunctionStmt.java]: And is function name case insensitive?

**Review by morningman — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/analysis/CreateFunctionStmt.java [hunk 3]
```diff
@@ -17,38 +17,129 @@
...
+    }
+
+    private void analyzeUda() throws AnalysisException {
+        throw new AnalysisException("Not support aggregate function now.");
+    }
+
+    private void analyzeUdf() throws AnalysisException {
+        final String SYMBOL_KEY = "symbol";
+        String symbol = properties.get(SYMBOL_KEY);
+        if (symbol == null) {
```
[the remark this answers is not in the document]
Comment (imay) [reply fe/src/main/java/org/apache/doris/analysis/CreateFunctionStmt.java]: OK，I'll change it

File: fe/src/main/java/org/apache/doris/analysis/CreateFunctionStmt.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (imay) [reply fe/src/main/java/org/apache/doris/analysis/CreateFunctionStmt.java]: > What if user create a function same as a builtin function? Do we allow it?

Now, I allow user create a same name UDF, but can not be used. I will add check later.


> And is function name case insensitive?

Yes, case insensitive

File: fe/src/main/java/org/apache/doris/analysis/CreateFunctionStmt.java [hunk 4]
```diff
@@ -17,38 +17,129 @@
...
+        }
+    }
+
+    private void analyzeCommon(Analyzer analyzer) throws AnalysisException {
+        // check function name
+        functionName.analyze(analyzer);
+
+        // check operation privilege
+        if (!Catalog.getCurrentCatalog().getAuth().checkDbPriv(
+                ConnectContext.get(), functionName.getDb(), PrivPredicate.ADMIN)) {
```
[the remark this answers is not in the document]
Comment (imay) [reply fe/src/main/java/org/apache/doris/analysis/CreateFunctionStmt.java]: > Use checkGlobalPriv()

OK

File: fe/src/main/java/org/apache/doris/catalog/Function.java (hunk 1 above)
Comment (imay) [reply fe/src/main/java/org/apache/doris/catalog/Function.java]: OK

**Review by morningman — APPROVED** [review]

**Pull request merged**